File: backend/app/tools/code_exec.py
"""Async code execution for Mogbot using subprocess."""

from __future__ import annotations

import asyncio


class CodeExecutor:
    """Execute code snippets in Python, bash, or JavaScript via subprocess."""

    def __init__(self, workdir: str) -> None:
        self.workdir = workdir

    async def execute(
        self, language: str, code: str, timeout: int = 30
    ) -> str:
        """Run code and return stdout+stderr (truncated).

        Args:
            language: One of 'python', 'bash', 'javascript'.
            code: The source code to execute.
            timeout: Maximum seconds before killing the process.

        Returns:
            Output string (max 10000 chars) or error string (max 5000 chars).
        """
        commands: dict[str, list[str]] = {
            "python": ["python3", "-c", code],
            "bash": ["bash", "-c", code],
            "javascript": ["node", "-e", code],
        }

        cmd = commands.get(language)
        if cmd is None:
            return f"Unsupported language: {language}"

        try:
            process = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                cwd=self.workdir,
            )

            try:
                stdout, stderr = await asyncio.wait_for(
                    process.communicate(), timeout=timeout
                )
            except asyncio.TimeoutError:
                process.kill()
                await process.wait()
                return f"Error: Execution timed out after {timeout} seconds"

            output = stdout.decode("utf-8", errors="replace")
            errors = stderr.decode("utf-8", errors="replace")

            if process.returncode != 0:
                combined = output + errors if output else errors
                return f"Error: {combined}"[:5000]

            return (output + errors)[:10000]

        except Exception as exc:
            return f"Error: {exc}"[:5000]

File: src/tools/code-exec.ts
import { execSync } from "child_process";
import * as fs from "fs";
import * as path from "path";
import * as os from "os";

export class CodeExecutor {
  private workdir: string;

  constructor(workdir: string) {
    this.workdir = workdir;
  }

  async execute(
    language: string,
    code: string,
    timeout = 30
  ): Promise<string> {
    const tmpFile = path.join(os.tmpdir(), `mogbot_${Date.now()}`);
    let cmd: string;

    switch (language) {
      case "python":
        fs.writeFileSync(`${tmpFile}.py`, code);
        cmd =
          `docker run --rm -v ${this.workdir}:/workspace -v ${tmpFile}.py:/tmp/script.py ` +
          `--network host python:3.11-slim python /tmp/script.py`;
        break;
      case "bash":
        fs.writeFileSync(`${tmpFile}.sh`, code);
        cmd =
          `docker run --rm -v ${this.workdir}:/workspace -v ${tmpFile}.sh:/tmp/script.sh ` +
          `--network host ubuntu:22.04 bash /tmp/script.sh`;
        break;
      case "javascript":
        fs.writeFileSync(`${tmpFile}.js`, code);
        cmd =
          `docker run --rm -v ${this.workdir}:/workspace -v ${tmpFile}.js:/tmp/script.js ` +
          `--network host node:20-slim node /tmp/script.js`;
        break;
      default:
        return `Unsupported language: ${language}`;
    }

    try {
      const output = execSync(cmd, {
        timeout: timeout * 1000,
        maxBuffer: 1024 * 1024,
        cwd: this.workdir
      });
      return output.toString().slice(0, 10000);
    } catch (err: any) {
      return `Error: ${err.stderr?.toString() || err.message}`.slice(0, 5000);
    }
  }
}

export class SimpleCodeExecutor {
  private workdir: string;

  constructor(workdir: string) {
    this.workdir = workdir;
  }

  async execute(
    language: string,
    code: string,
    timeout = 30
  ): Promise<string> {
    const cmds: Record<string, string[]> = {
      python: ["python3", "-c", code],
      bash: ["bash", "-c", code],
      javascript: ["node", "-e", code]
    };

    const args = cmds[language];
    if (!args) return `Unsupported: ${language}`;

    try {
      const output = execSync(args.join(" "), {
        timeout: timeout * 1000,
        cwd: this.workdir,
        maxBuffer: 1024 * 1024
      });
      return output.toString().slice(0, 10000);
    } catch (err: any) {
      return `Error: ${err.stderr?.toString() || err.message}`.slice(0, 5000);
    }
  }
}
